Use np.digitize for segmentation.relabel_sequential

ajaybhat opened the pull request:
Fix for #1349 
Should <code>label_field</code> be renamed to <code>image</code> as in the issue?

JDWarner commented:
Ignore the AppVeyor failure, that's unrelated. The 7 Travis test failures seem to be along the lines of integer offsets. I'm not sure exactly what's going on, but if you want to work on this locally you can run just the `relabel` tests by entering the line 

``` python
run skimage/segmentation/tests/test_join.py
```

into an IPython console running in the package root. You can also inspect [the tests directly](https://github.com/scikit-image/scikit-image/blob/master/skimage/segmentation/tests/test_join.py) to reproduce and dig deeper.

jni commented on skimage/segmentation/_join.py:
```diff
@@ -120,14 +120,11 @@ def relabel_sequential(label_field, offset=1):
-    labels0 = labels[labels != 0]
-    if m == len(labels0):  # nothing to do, already 1...n labels
-        return label_field, labels, labels
+    relabeled = np.digitize(label_field.reshape(-1, ), labels).reshape(label_field.shape)
```
In addition to addressing the test failures, change `reshape(-1, )` to `ravel()`

ajaybhat commented:
I've fixed the test errors and changed the code as per advice by @jni 
Since offset should be strictly positive I also raise an exception if a value not positive is passed

ajaybhat commented:
Is there anything else to change?

jni commented:
@ajaybhat I'm just wondering whether there's a way to simplify this. This adds 5 lines to the codebase for not much benefit that I can see. Is this version substantially faster, for example?

(Although, does the update to `inverse_map` fix a bug? If so, that should be tested to prevent a regression.)

ajaybhat commented:
> I'm just wondering whether there's a way to simplify this. This adds 5 lines to the codebase for not much benefit that I can see.

I've rewritten some of it to reduce the lines of code.

> (Although, does the update to inverse_map fix a bug? 

Not really. I just saw an opportunity to use <code>np.concatenate</code>

ajaybhat commented:
@jni > I'm just wondering whether there's a way to simplify this. This adds 5 lines to the codebase for not much benefit that I can see.

I've rewritten some of it to reduce the lines of code.

> (Although, does the update to inverse_map fix a bug? 

Not really. I just saw an opportunity to use <code>np.concatenate</code>

jni commented:
@ajaybhat 

> Not really. I just saw an opportunity to use `np.concatenate`

I meant this line:

```python
inverse_map = np.concatenate(([0], inverse_map)) if has_zero else inverse_map
```

I didn't add this before. Was that a bug?

Also, even after your simplification, I still think the original code was clearer. (I may be biased though — maybe someone else can weigh in, @JDWarner @stefanv?) Certain parts of the new code, I suspect, might also be slower, for example:

```python
relabeled = np.array([np.add(labeled, offset - 1) for labeled in relabeled if
                      (labeled.ndim == 0 and labeled != 0) or labeled.ndim != 0])
```

This is a pure-Python loop as long as the image. The name `labeled` is confusing here. And I don't understand the conditions.

jni commented:
@ajaybhat it seems we dropped the ball on this... Do you have time to answer my questions above? 

@JDWarner @stefanv @ahojnnes any opinion on whether this code is clearer than the current version in master?

souravsingh commented:
@jni Is the PR ready for merging?

jni commented:
@souravsingh No, I would say it's ready for closing, since the original author has not responded... Unless someone wants to run benchmarks that show that this is substantially faster/more memory-efficient.

jni commented:
Ok so I've run some timings, and in my testing, this version is almost 3x version than the existing function:

```python
In [3]: labels = np.random.randint(0, 10000, size=(1024, 1024))

In [4]: %timeit relabel_sequential(labels)  # as defined in #1531
1 loop, best of 3: 279 ms per loop

In [5]: from skimage import segmentation as seg

In [6]: %timeit seg.relabel_sequential(labels)
10 loops, best of 3: 99.8 ms per loop
```

Therefore I'm going to close this. As usual, feel free to reopen if there are new improvements.
